sematic/user_settings.py

```diff
@@ -88,57 +46,326 @@ class SettingsVar(enum.Enum):
-        )
+"""
+        ).format(missing_settings.value, missing_settings.value)
+
         super().__init__(message)
 
 
-def get_user_settings(var: SettingsVar, *args) -> str:
+@dataclass
+class UserSettings:
     """
-    Main API to access individual settings.
-
-    Default as first args or will raise if no settings.
+    The representation of the user's settings.
+
+    Parameters
+    ----------
+    active_profile: str
+        The profile whose settings will actually be used
+    profiles: Dict[str, ProfileSettingsType]
+        The settings for each defined profile
     """
-    if var not in SettingsVar.__members__.values():
-        raise ValueError(
-            "Invalid settings var: {}. Available vars: {}".format(
-                repr(var), SettingsVar.__members__.values()
+
+    active_profile: str
+    profiles: Dict[str, ProfileSettingsType]
+
+    def __post_init__(self):
+        """
+        Validates the consistency of profile data.
+
+        Raises
+        ------
+        ValueError:
+            There is an incorrect value or state
+        """
+        if len(self.active_profile) == 0:
+            raise ValueError("The active profile cannot be empty!")
+
+        if len(self.profiles) == 0:
+            # be lenient in what you accept, so just fill it in and carry on
+            self.profiles = {self.active_profile: {}}
+        else:
+            if self.active_profile not in self.profiles:
+                raise ValueError(
+                    f"The '{self.active_profile}' active profile is not present in the "
+                    f"listed profiles: {list(self.profiles.keys())}"
+                )
+
+        # impose the enums over the settings values
+        profiles = list(self.profiles.keys())
+        for profile in profiles:
+            self.profiles[profile] = {
+                SettingsVar[var]: value  # type: ignore
+                for var, value in self.profiles[profile].items()
+            }
+
+    def get_active_profile_settings(self) -> ProfileSettingsType:
+        """
+        Returns the settings associated with the active profile.
+        """
+        return self.profiles[self.active_profile]
+
+    def get_inactive_profiles(self) -> List[str]:
+        """
+        Returns a list of existing profile names other than the active one.
+        """
+        return [
+            profile
+            for profile in self.profiles.keys()
+            if profile != self.active_profile
+        ]
+
+    def set(self, var: SettingsVar, value: Optional[str]) -> None:
+        """
+        Sets the specified value for the specified variable, for the active profile.
+
+        Deletes the entry if the value is None.
+        """
+        active_profile_settings = self.get_active_profile_settings()
+
+        if value is not None:
+            active_profile_settings[var] = value
+            return
+
+        if var not in active_profile_settings:
```

Comment:
when would this happen?

Reply:
This was for the delete branch of the set method, which happened when the value to set was None. I split out a separate delete method, as this was hard to understand.